File: problem7.c
#include <stdio.h>

int is_prime(int n, int primes[]) {
    int i;
    for (i = 0; i < 10001; i++) {
        if (primes[i] == 0)
            break;
        if (n % primes[i] == 0)
            return 0;
    }
    return 1;
}

int main() {
    int i = 0;  // primes index
    int n = 2;
    int primes[10001] = {0};  // initialize array to zero

    while (i < 10001) {
        if (is_prime(n, primes)) {
            primes[i] = n;
            i++;
        }
        n++;
    }

    printf("%d\n", primes[10000]);
    return 0;
}

File: problem7.py

def is_prime(n, primes):
    for p in primes:
        if n % p == 0:
            return False
    return True

i = 2
primes = []
while True:
    if len(primes) >= 10001:
        break
    if is_prime(i, primes):
        primes.append(i)
    i += 1
print(primes[10000])
